source/mir/combinatorics/package.d

```diff
@@ -0,0 +1,1109 @@
+    in
+    {
+        assert(state.length + 1 == indices.length);
+    }
+    body
+    {
+        // iota
+        foreach(uint i, ref index; indices)
+            index = i;
+        state[] = 0;
+
+        this.indices = indices;
+        this.state = state;
+
+        _empty = indices.length == 0;
+    }
+
+    @property uint[] front() @safe pure nothrow @nogc
+    {
+        return indices;
+    }
+
+    void popFront() @safe pure nothrow @nogc
+    {
+        import std.algorithm.mutation : swapAt;
+        for (uint h = 0;;h++)
+        {
+            if (h+2 > indices.length)
+            {
+                _empty = true;
+                break;
+            }
+
+            if (h & 1)
+                indices.swapAt(0, h+1);
+            else
+                indices.swapAt(state[h], h+1);
+
+            if (state[h] == h+1)
+            {
+                state[h] = 0;
+                continue;
+            }
+            state[h]++;
+            break;
+        }
+    }
+
+    @property bool empty() @safe pure nothrow @nogc
+    {
+        return _empty;
+    }
+
+    @property Permutations save() @safe pure nothrow
+    {
+        typeof(this) c = this;
+        c.indices = indices.dup;
+        c.state = state.dup;
+        return c;
+    }
+}
+
+/**
+Lazily computes the Cartesian power of $(D r) with itself
+for a number of repetitions $(D repeat).
+If the input is sorted, the product is in lexicographic order.
+
+While generating a new item is in `O(k)` (amortized `O(1)`),
+the total number of elements is `n^k`.
+
+For example $(D"AB".cartesianPower(2).array) returns $(D["AA", "AB", "BA", "BB"])
+
+Params:
+    n = number of elements (`|r|`)
+    r = $(REF_ALTTEXT Input range, isInputRange, std, range, primitives)
+    repeat = number of repetitions
+    alloc = custom Allocator
+
+Returns:
+    Forward range which yields the product items
+*/
+CartesianPower cartesianPower(size_t n, size_t repeat = 1) @safe pure nothrow
+in
+{
+    assert(repeat >= 1, "Invalid number of repetitions");
+}
+body
+{
+    return CartesianPower(n, new uint[repeat]);
+}
+
+/// ditto
+IndexedRoR!(CartesianPower, Range) cartesianPower(Range)(Range r, size_t repeat = 1)
+if (isInputRange!Range && hasLength!Range)
+in
+{
+    assert(repeat >= 1, "Invalid number of repetitions");
+}
+body
+{
+    return cartesianPower(r.length, repeat).indexedRoR(r);
+}
+
+/// ditto
+CartesianPower makeCartesianPower(Allocator)(size_t n, size_t repeat, auto ref Allocator alloc)
+in
+{
+    assert(repeat >= 1, "Invalid number of repetitions");
+}
+body
+{
+    import std.experimental.allocator: makeArray;
+    return CartesianPower(n, alloc.makeArray!uint(repeat));
+}
+
+/// Dispose a CartesianPower object
+void dispose(Allocator)(auto ref Allocator alloc, auto ref CartesianPower cartesianPower)
+{
+    import std.experimental.allocator: dispose;
+    dispose(alloc, cartesianPower._state);
+}
+
+private struct CartesianPower
```

Comment:
BTW, this struct should be private/package. Just undocumented. Otherwise user would not be able to access methods